File: app/src/main/java/org/tvheadend/tvhclient/data/dao/ChannelTagDao.java
package org.tvheadend.tvhclient.data.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import org.tvheadend.tvhclient.domain.entity.ChannelTag;

import java.util.List;

@Dao
public interface ChannelTagDao {

    String CONNECTION_IS_ACTIVE = " connection_id IN (SELECT id FROM connections WHERE active = 1) ";

    @Query("SELECT DISTINCT * FROM channel_tags " +
            "WHERE " + CONNECTION_IS_ACTIVE +
            "ORDER BY tag_name")
    LiveData<List<ChannelTag>> loadAllChannelTags();

    @Query("SELECT DISTINCT * FROM channel_tags " +
            "WHERE " + CONNECTION_IS_ACTIVE +
            "ORDER BY tag_name")
    List<ChannelTag> loadAllChannelTagsSync();

    @Query("SELECT DISTINCT * FROM channel_tags " +
            "WHERE " + CONNECTION_IS_ACTIVE +
            "AND id = :id ")
    ChannelTag loadChannelTagByIdSync(int id);

    @Transaction
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(ChannelTag channelTag);

    @Transaction
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(List<ChannelTag> items);

    @Update
    void update(List<ChannelTag> channelTags);

    @Update
    void update(ChannelTag channelTags);

    @Delete
    void delete(ChannelTag channelTag);

    @Query("DELETE FROM channel_tags")
    void deleteAll();

    @Query("DELETE FROM channel_tags " +
            "WHERE " + CONNECTION_IS_ACTIVE +
            "AND id = :id ")
    void deleteById(int id);

    @Query("SELECT id FROM channel_tags " +
            "WHERE " + CONNECTION_IS_ACTIVE +
            "AND is_selected = 1 " +
            "ORDER BY tag_name")
    LiveData<List<Integer>> loadAllSelectedItemIds();

    @Query("SELECT COUNT (*) FROM channel_tags " +
            "WHERE " + CONNECTION_IS_ACTIVE)
    int getItemCountSync();
}

File: app/src/main/java/org/tvheadend/tvhclient/data/dao/ChannelTagDao.kt
package org.tvheadend.tvhclient.data.dao

import androidx.lifecycle.LiveData
import androidx.room.*
import org.tvheadend.tvhclient.domain.entity.ChannelTag

@Dao
interface ChannelTagDao {

    @get:Query("SELECT COUNT (*) FROM channel_tags " +
            " WHERE $CONNECTION_IS_ACTIVE")
    val itemCountSync: Int

    @Query("SELECT DISTINCT * FROM channel_tags " +
            " WHERE $CONNECTION_IS_ACTIVE" +
            " ORDER BY tag_name")
    fun loadAllChannelTags(): LiveData<List<ChannelTag>>

    @Query("SELECT DISTINCT * FROM channel_tags " +
            " WHERE $CONNECTION_IS_ACTIVE" +
            " ORDER BY tag_name")
    fun loadAllChannelTagsSync(): List<ChannelTag>

    @Query("SELECT DISTINCT * FROM channel_tags " +
            " WHERE $CONNECTION_IS_ACTIVE" +
            " AND id = :id ")
    fun loadChannelTagByIdSync(id: Int): ChannelTag

    @Transaction
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insert(channelTag: ChannelTag)

    @Transaction
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insert(items: List<ChannelTag>)

    @Update
    fun update(channelTags: List<ChannelTag>)

    @Update
    fun update(channelTags: ChannelTag)

    @Delete
    fun delete(channelTag: ChannelTag)

    @Query("DELETE FROM channel_tags")
    fun deleteAll()

    @Query("DELETE FROM channel_tags " +
            " WHERE $CONNECTION_IS_ACTIVE" +
            " AND id = :id ")
    fun deleteById(id: Int)

    @Query("SELECT id FROM channel_tags " +
            " WHERE $CONNECTION_IS_ACTIVE" +
            " AND is_selected = 1 " +
            " ORDER BY tag_name")
    fun loadAllSelectedItemIds(): LiveData<List<Int>?>

    companion object {

        const val CONNECTION_IS_ACTIVE = " connection_id IN (SELECT id FROM connections WHERE active = 1) "
    }
}
